src/fetch/rewards.py

```diff
@@ -0,0 +1,170 @@
+    )
+    secondary_rewards_df = (
+        batch_data_df["participating_solvers"]
+        .explode()
+        .value_counts()
+        .reset_index()
+        .rename(columns={"participating_solvers": "solver", "count": "num_auctions"})
+        .astype(object)
+    )
+    participation_total = secondary_rewards_df["num_auctions"].sum()
+    secondary_rewards_df["secondary_reward_cow"] = (
+        secondary_allocation * secondary_rewards_df["num_auctions"]
+    ) // participation_total
+    secondary_rewards_df["secondary_reward_eth"] = (
+        secondary_rewards_df["secondary_reward_cow"]
+        .apply(converter.token_to_eth)
+        .astype(object)
+    )
+    secondary_rewards_df = secondary_rewards_df.drop("num_auctions", axis=1)
+
+    return secondary_rewards_df
+
+
+def compute_quote_rewards(
+    trade_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute quote rewards from batch data"""
+    quote_rewards_df = (
+        trade_data_df[trade_data_df.apply(is_market_order, axis=1)]
+        .quote_solver.value_counts()
+        .reset_index()
+        .rename(columns={"quote_solver": "solver", "count": "num_quotes"})
+        .apply(
+            lambda row: (
+                row["solver"],
+                int(
+                    float(
+                        min(
+                            QUOTE_REWARD_COW,
+                            converter.eth_to_token(QUOTE_REWARD_CAP_ETH),
+                        )
+                    )
+                    * row["num_quotes"]
+                ),
+            ),
+            axis=1,
+            result_type="expand",
+        )
+        .rename(columns={0: "solver", 1: "quote_reward_cow"})
+        .astype(object)
+    )
+    return quote_rewards_df
+
+
+def is_market_order(row: Series) -> bool:
+    """Check if an order was in market when created."""
+    try:
+        if row["quote_solver"] is None:
+            return False
+        if row["kind"] == "sell":
+            return (
+                int(
+                    row["quote_sell_amount"]
+                    - row["quote_gas_amount"]
+                    * row["quote_gas_price"]
+                    / row["quote_sell_token_price"]
+                )
+                * row["quote_buy_amount"]
+                >= row["limit_buy_amount"] * row["quote_sell_amount"]
+                and not row["partially_fillable"]
+            )
+        if row["kind"] == "buy":
+            return (
+                row["limit_sell_amount"]
+                >= int(
+                    row["quote_sell_amount"]
+                    + row["quote_gas_amount"]
+                    * row["quote_gas_price"]
+                    / row["quote_sell_token_price"]
+                )
+                and not row["partially_fillable"]
+            )
+        raise ValueError(
+            f"Unknow order kind \"{row['kind']}\". Only \"sell\" and \"buy\" are supported."
+        )
+    except KeyError:
+        return False
```

Comment:
Has this key error ever happened? If so, do we know why? Might be worth commenting when a key error could occur.

Reply:
Nice catch. I had those errors with some earlier version of the code only. With the current code, the sql query always returns those columns and a key error should never happen.